evm-bridging/cadence/transactions/wrap_nfts.cdc

```diff
@@ -0,0 +1,89 @@
+import "EVM"
+
+/// Wraps NFTs with provided IDs
+///
+/// @param wrapperERC721Address: EVM address of the wrapper ERC721 NFT
+/// @param nftIDs: Array of IDs of the NFTs to wrap
+///
+transaction(
+    wrapperERC721Address: String,
+    nftIDs: [UInt256]
+) {
+    // Cadence-owned account
+    let coa: auth(EVM.Call) &EVM.CadenceOwnedAccount
+
+    prepare(signer: auth(BorrowValue) &Account) {
+        // Borrow a reference to the signer's COA
+        self.coa = signer.storage.borrow<auth(EVM.Call) &EVM.CadenceOwnedAccount>(from: /storage/evm)
+            ?? panic("No COA found in signer's account")
+    }
+
+    execute {
+        // Get contract addresses
+        let wrapperAddress = EVM.addressFromString(wrapperERC721Address)
+        let underlyingAddress = getUnderlyingERC721Address(self.coa, wrapperAddress)
+
+        // Approve contract to withdraw underlying NFTs from signer's coa
+        mustCall(self.coa, underlyingAddress,
+            functionSig: "setApprovalForAll(address,bool)",
+            args: [wrapperAddress, true]
+        )
+
+        // Wrap NFTs with provided IDs
+        mustCall(self.coa, wrapperAddress,
+            functionSig: "depositFor(address,uint256[])",
+            args: [self.coa.address(), nftIDs]
+        )
+
+        // Revoke approval for contract to withdraw underlying NFTs from signer's coa
+        mustCall(self.coa, underlyingAddress,
+            functionSig: "setApprovalForAll(address,bool)",
+            args: [wrapperAddress, false]
+        )
+    }
+}
+
+/// Gets the underlying ERC721 address
+///
+access(all) fun getUnderlyingERC721Address(
+    _ coa: auth(EVM.Call) &EVM.CadenceOwnedAccount,
+    _ wrapperAddress: EVM.EVMAddress
+): EVM.EVMAddress {
+    let res = coa.call(
+        to: wrapperAddress,
+        data: EVM.encodeABIWithSignature("underlying()", []),
+        gasLimit: 100_000,
+        value: EVM.Balance(attoflow: 0)
+    )
+
+    assert(res.status == EVM.Status.successful, message: "Call to get underlying ERC721 address failed")
+    let decodedResult = EVM.decodeABI(types: [Type<EVM.EVMAddress>()], data: res.data)
+    assert(decodedResult.length == 1, message: "Invalid response length")
+
+    return decodedResult[0] as! EVM.EVMAddress
+}
+
+/// Calls a function on an EVM contract from provided coa
```

Comment:
so we decided not to deploy an util contract for the `mustCall` function?

Reply:
Yea it would be best achieved by holding the contract in a trusted neutral entity like the bridge account or a keyless account given the privileged access to COA, but it doesn't seem really worth it at this point as the wrapping/unwrapping logic is needed only temporarily until the bridge gains support for custom associations and unwrapping logic within its own context